File: libraries/go/crypto.go
package vsync

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"crypto/sha256"
	"fmt"

	"golang.org/x/crypto/pbkdf2"
)

// RQE1 envelope (v0.2 §3):
//
//	bytes 0..3    magic "RQE1"
//	bytes 4..15   12-byte IV
//	bytes 16..N   ciphertext || 16-byte AES-GCM tag
//
// KDF: PBKDF2-HMAC-SHA256 → 32-byte AES-256 key.
const (
	rqe1HeaderLen = 16 // 4-byte magic + 12-byte IV
	rqe1GCMTagLen = 16
	rqe1KeyLen    = 32
	rqe1IVLen     = 12
	// Structural minimum: header (magic + IV) + GCM tag is the smallest a
	// valid envelope can be — that's the empty-plaintext case. Anything
	// shorter is mid-payload truncation we surface as BundleCorrupt rather
	// than letting AES-GCM raise the ambiguous InvalidTag (which we'd have
	// to surface as WrongPassphrase). The conformance corpus's truncated-
	// ciphertext vector is 30 bytes and fires this branch; the empty-
	// plaintext positive vector is exactly 32 bytes and passes through.
	// The team-lead's brief listed 48 (counting salt as part of the
	// envelope) — but salt is a separate PBKDF2 input in the RQE1 wire
	// format, so 32 is the honest floor.
	rqe1MinEnvelope   = rqe1HeaderLen + rqe1GCMTagLen // 32
	defaultPBKDF2Iter = 600_000
)

var (
	rqe1MagicPrefix = []byte{'R', 'Q', 'E'}
	rqe1Version     = byte('1')
)

// DecryptRQE1 decrypts an RQE1 envelope. The salt is the UTF-8 bytes of the
// caller-supplied string (matches the CLI's PBKDF2 path and the conformance
// rqe1-decrypt vectors, which carry salt as a base64-ish ASCII string fed
// to PBKDF2 verbatim).
//
// Error mapping (v0.12 §11):
//   - envelope shorter than rqe1MinEnvelope    → ErrBundleCorrupt
//   - magic prefix "RQE" missing                → ErrBundleCorrupt
//   - magic prefix ok but version byte != '1'   → ErrUnsupportedSpecVersion
//   - GCM tag rejects                           → ErrWrongPassphrase
func DecryptRQE1(envelope []byte, passphrase, salt string, iterations int) ([]byte, error) {
	return decryptRQE1WithSaltBytes(envelope, passphrase, []byte(salt), iterations)
}

// decryptRQE1WithSaltBytes is the byte-flavored entry point used by the
// config-blob path (which decodes a standard-base64 salt field into raw
// bytes). Keeps the runtime call site honest about whether it's feeding
// a string or pre-decoded bytes to PBKDF2.
func decryptRQE1WithSaltBytes(envelope []byte, passphrase string, salt []byte, iterations int) ([]byte, error) {
	// Structural-length gate — see the team-lead's heuristic note:
	// AES-GCM raises the same InvalidTag for "tag flipped" and "ciphertext
	// clipped mid-payload"; the only honest way to disambiguate without an
	// explicit length field is structural. < 48 bytes is impossible for a
	// valid envelope (magic+salt+IV+tag minimum the spec discussions pinned).
	if len(envelope) < rqe1MinEnvelope {
		return nil, fmt.Errorf("%w: RQE1 envelope structurally too short (%d bytes < %d minimum)",
			ErrBundleCorrupt, len(envelope), rqe1MinEnvelope)
	}
	if len(envelope) < rqe1HeaderLen+rqe1GCMTagLen {
		// Defensive — covered by the floor above but explicit for the reader.
		return nil, fmt.Errorf("%w: RQE1 envelope too short for header+tag", ErrBundleCorrupt)
	}
	// Split the 4-byte magic into prefix + version so wrong-version is
	// reported as UnsupportedSpecVersion (v0.12 §11) rather than corrupt.
	if !equalBytes(envelope[:3], rqe1MagicPrefix) {
		return nil, fmt.Errorf("%w: RQE1 magic prefix is not 'RQE'", ErrBundleCorrupt)
	}
	if envelope[3] != rqe1Version {
		return nil, fmt.Errorf("%w: RQE envelope advertises version %q; this library understands '1' only",
			ErrUnsupportedSpecVersion, envelope[3:4])
	}

	iv := envelope[4:rqe1HeaderLen]
	ciphertext := envelope[rqe1HeaderLen:]

	key := pbkdf2.Key([]byte(passphrase), salt, iterations, rqe1KeyLen, sha256.New)
	block, err := aes.NewCipher(key)
	if err != nil {
		// rqe1KeyLen is constant 32 — this can only fail in a broken build.
		return nil, fmt.Errorf("%w: aes init: %v", ErrBundleCorrupt, err)
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, fmt.Errorf("%w: gcm init: %v", ErrBundleCorrupt, err)
	}
	if len(iv) != gcm.NonceSize() {
		return nil, fmt.Errorf("%w: IV size mismatch", ErrBundleCorrupt)
	}
	pt, err := gcm.Open(nil, iv, ciphertext, nil)
	if err != nil {
		return nil, fmt.Errorf("%w: AES-GCM tag rejected — passphrase wrong or ciphertext tampered (%v)",
			ErrWrongPassphrase, err)
	}
	return pt, nil
}

// encryptRQE1ForTest is unexported — production callers never encrypt; the
// CLI is the canonical writer. The unit suite uses this to mint round-trip
// fixtures independent of docs/specs/test-vectors/.
func encryptRQE1ForTest(plaintext []byte, passphrase, salt string, iterations int) ([]byte, error) {
	iv := make([]byte, rqe1IVLen)
	if _, err := rand.Read(iv); err != nil {
		return nil, err
	}
	key := pbkdf2.Key([]byte(passphrase), []byte(salt), iterations, rqe1KeyLen, sha256.New)
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, err
	}
	ct := gcm.Seal(nil, iv, plaintext, nil)
	out := make([]byte, 0, rqe1HeaderLen+len(ct))
	out = append(out, rqe1MagicPrefix...)
	out = append(out, rqe1Version)
	out = append(out, iv...)
	out = append(out, ct...)
	return out, nil
}

func equalBytes(a, b []byte) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

File: libraries/typescript/src/crypto.ts
// RQE1 envelope — decrypt path only. (Read-side library; the CLI is the
// canonical writer.)
//
// Envelope layout (mirrors `src/crypto.ts` in the CLI):
//   bytes 0..3    magic "RQE1"
//   bytes 4..15   12-byte IV (random per encryption)
//   bytes 16..N   ciphertext || 16-byte AES-GCM auth tag
//
// KDF: PBKDF2-HMAC-SHA256, 600_000 iterations → 32-byte AES-256 key.
//
// ─── Salt byte-semantics (load-bearing) ─────────────────────────────────
//
// Per v0.12 §2.1 (post-revision) and the Python reference implementation,
// the `salt` field is a STRING and its **UTF-8 bytes** are the input to
// PBKDF2. Readers MUST NOT base64-decode the salt before feeding it to
// PBKDF2 — the field's base64-like alphabet is a CLI storage artefact
// (the CLI's per-(repo, env) salt is generated as `b64nopad(random_18b)`
// and stored as that ASCII string in the on-disk config), not a wire
// transport-encoding marker.
//
// This convention is what:
//   • the CLI's `src/crypto.ts::deriveKey` does (calls `enc.encode(salt)`)
//   • the test-vector generator does (encrypts via `encryptWithIV(pt, pw, saltStr, iv)`)
//   • the Python reference's `crypto._derive_key` does (`salt.encode("utf-8")`)
//
// Spec contract: feed the UTF-8 bytes of the string directly. Library
// `BundleCorruptError(< 48 bytes structural truncation)` heuristic per
// Wave 6 brief.

import { createDecipheriv, createCipheriv, pbkdf2Sync, randomBytes } from "node:crypto";
import {
  BundleCorruptError,
  UnsupportedSpecVersionError,
  WrongPassphraseError,
} from "./errors.js";

const MAGIC_PREFIX = Buffer.from("RQE", "ascii"); // 3 bytes
const MAGIC_VERSION = Buffer.from("1", "ascii"); // 1 byte
const MAGIC = Buffer.concat([MAGIC_PREFIX, MAGIC_VERSION]); // "RQE1"
const IV_LEN = 12;
const TAG_LEN = 16;
const HEADER_LEN = MAGIC.length + IV_LEN; // 16
const KEY_LEN = 32;
export const PBKDF2_ITERATIONS = 600_000;
// Structural truncation floor — magic(4) + IV(12) + ciphertext≥0 +
// tag(16) = 32-byte minimum for a syntactically valid RQE1 envelope.
// Anything below 32 bytes can't structurally hold even an empty
// ciphertext + auth tag, so we classify as BundleCorruptError before
// PBKDF2 — this disambiguates the v0.11 `truncated-ciphertext` (30 bytes)
// vector from the `wrong-passphrase` / `bad-gcm-tag` cases (≥32 bytes,
// AES-GCM tag failure → WrongPassphraseError).
//
// Matches Python's `HEADER_LEN(16) + MIN_CT_LEN(16) = 32` floor.
const MIN_ENVELOPE_BYTES = 32;

function deriveKey(passphrase: string, saltString: string, iterations: number): Buffer {
  return pbkdf2Sync(
    Buffer.from(passphrase, "utf8"),
    Buffer.from(saltString, "utf8"),
    iterations,
    KEY_LEN,
    "sha256",
  );
}

/**
 * Decrypt an RQE1 envelope. Maps crypto failures to the v0.12 taxonomy:
 *   - too-short / corrupt magic prefix / structural truncation → BundleCorruptError
 *   - magic prefix "RQE" but version byte != "1"               → UnsupportedSpecVersionError
 *   - GCM tag rejects (wrong passphrase or tampering)          → WrongPassphraseError
 *
 * `salt` is the STRING from the config blob; UTF-8 bytes are fed to PBKDF2.
 */
export async function decryptRqe1(
  blob: Uint8Array,
  passphrase: string,
  salt: string,
  iterations: number = PBKDF2_ITERATIONS,
): Promise<Uint8Array> {
  if (blob.byteLength < MIN_ENVELOPE_BYTES) {
    throw new BundleCorruptError(
      `RQE1 envelope too short: ${blob.byteLength} bytes (need at least ${MIN_ENVELOPE_BYTES} for magic+IV+tag)`,
    );
  }
  const buf = Buffer.from(blob.buffer, blob.byteOffset, blob.byteLength);
  if (!buf.subarray(0, MAGIC_PREFIX.length).equals(MAGIC_PREFIX)) {
    throw new BundleCorruptError(
      "RQE1 envelope: magic prefix is not 'RQE' — not a vsync envelope",
    );
  }
  if (!buf.subarray(MAGIC_PREFIX.length, MAGIC.length).equals(MAGIC_VERSION)) {
    const got = buf.subarray(MAGIC_PREFIX.length, MAGIC.length).toString("ascii");
    throw new UnsupportedSpecVersionError(
      `RQE envelope advertises version '${got}'; this library understands version '1' only — upgrade vsync-s3-client`,
    );
  }

  const iv = buf.subarray(MAGIC.length, HEADER_LEN);
  const ctAndTag = buf.subarray(HEADER_LEN);
  // Node's createDecipheriv wants ct and tag separately; AES-GCM tag is
  // the last 16 bytes by convention in our envelope.
  const ct = ctAndTag.subarray(0, ctAndTag.byteLength - TAG_LEN);
  const tag = ctAndTag.subarray(ctAndTag.byteLength - TAG_LEN);

  const key = deriveKey(passphrase, salt, iterations);
  const decipher = createDecipheriv("aes-256-gcm", key, iv);
  decipher.setAuthTag(tag);
  try {
    const pt = Buffer.concat([decipher.update(ct), decipher.final()]);
    return new Uint8Array(pt.buffer, pt.byteOffset, pt.byteLength);
  } catch (e) {
    // Node's AES-GCM raises a plain Error with message containing
    // "Unsupported state or unable to authenticate data" on tag rejection.
    // We collapse to WrongPassphraseError per the taxonomy.
    throw new WrongPassphraseError(
      `RQE1 envelope: AES-GCM tag rejected — the passphrase is wrong or the ciphertext has been tampered with (${(e as Error).message})`,
    );
  }
}

/**
 * Round-trip helper for the unit suite — do NOT call from production.
 * Production encryption belongs to the CLI; this helper exists so the
 * test suite can mint envelopes without depending on the corpus.
 */
export async function encryptRqe1ForTest(
  plaintext: Uint8Array,
  passphrase: string,
  salt: string,
  iterations: number = PBKDF2_ITERATIONS,
): Promise<Uint8Array> {
  const iv = randomBytes(IV_LEN);
  const key = deriveKey(passphrase, salt, iterations);
  const cipher = createCipheriv("aes-256-gcm", key, iv);
  const ct = Buffer.concat([cipher.update(plaintext), cipher.final()]);
  const tag = cipher.getAuthTag();
  const out = Buffer.concat([MAGIC, iv, ct, tag]);
  return new Uint8Array(out.buffer, out.byteOffset, out.byteLength);
}
